### PR TITLE
*: use OpenSSL for crypto RNG

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: ref #15982

What's Changed:

```commit-message
To comply with FIPS 140-2 requirements, it's essential to choose an RNG
that meets these specifications.
This commit replaces the `rand` crate with OpenSSL for cryptographic random
number generation.
```

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: components/encryption/src/io.rs [hunk 1]
```diff
@@ -741,13 +734,13 @@ mod tests {
...
         let mut plaintext = vec![0; 10240];
-        OsRng.fill_bytes(&mut plaintext);
+        rand::rand_bytes(&mut plaintext).unwrap();
         let offsets = [1024, 1024 + 1, 10240 - 1];
         let sizes = [1024, 8000];
         let written = vec![0; 10240];
         for method in methods {
-            let key = generate_data_key(method);
-            let iv = Iv::new_ctr();
+            let (_, key) = generate_data_key(method).unwrap();
```
Comment (LykxSassinator) [inline components/encryption/src/io.rs:742]: The error can be ignored ? 
Can we think the related `rand` function in `openssl::rand` always generate a bunch of random bytes as expected ?

**Review by LykxSassinator — COMMENTED** [review]
Rest LGTM

File: components/encryption/src/io.rs (hunk 1 above)
Comment (overvenus) [reply components/encryption/src/io.rs:742]: It should not be ignored, as an error meaning the OpenSSL CSPRNG can not generate secure random numbers.
The error is rare on recent platform.

File: components/encryption/Cargo.toml [hunk 2]
```diff
@@ -30,7 +31,9 @@ online_config = { workspace = true }
 openssl = { workspace = true }
 prometheus = { version = "0.13", features = ["nightly"] }
 protobuf = { version = "2.8", features = ["bytes"] }
-rand = "0.8"
+# For simplicity, do not use rand in ecnryption, as FIPs 140 has requirements
+# to random number generation.
```
Comment (overvenus) [inline components/encryption/Cargo.toml:35]:
```suggestion
# For simplicity and compliance with FIPS 140 requirements for random number
# generation, do not use the 'rand' crate in encryption-related code.
```

File: components/crypto/src/rand.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+pub fn rand_u64() -> Result<u64, ErrorStack> {
+    let mut rand_id = [0u8; 8];
+    rand::rand_bytes(&mut rand_id)?;
+    Ok(u64::from_ne_bytes(rand_id))
+}
+
+/// Fill buffer with cryptographically strong pseudo-random bytes.
+pub fn rand_bytes(buf: &mut [u8]) -> Result<(), ErrorStack> {
+    rand::rand_bytes(buf)
+}
```
Comment (LykxSassinator) [inline components/crypto/src/rand.rs:17]:
```suggestion
/// Fill buffer with cryptographically strong pseudo-random bytes.
pub fn rand_bytes(buf: &mut [u8]) -> Result<(), ErrorStack> {
    rand::rand_bytes(buf)
}

/// Return a random u64.
pub fn rand_u64() -> Result<u64, ErrorStack> {
    let mut rand_id = [0u8; 8];
    rand_bytes(&mut rand_id)?;
    Ok(u64::from_ne_bytes(rand_id))
}
```

Comment (LykxSassinator) [reply components/crypto/src/rand.rs:17]: Why not directly use `crate::rand_bytes()` in the implementation of `rand_u64()` to unify the calling of generating random bytes?

File: components/crypto/Cargo.toml [hunk 4]
```diff
@@ -1,5 +1,5 @@
 [package]
-name = "fips"
+name = "crypto"
```
Comment (Connor1996) [inline components/crypto/Cargo.toml:2]: why rename

Comment (overvenus) [reply components/crypto/Cargo.toml:2]: The crates contains not only FIPS utilities, but also crypto operations (RNG). 
See doc comments in lib.rs

File: components/crypto/src/rand.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+use openssl::{error::ErrorStack, rand};
+
+/// Fill buffer with cryptographically strong pseudo-random bytes.
+pub fn rand_bytes(buf: &mut [u8]) -> Result<(), ErrorStack> {
+    rand::rand_bytes(buf)
+}
+
+/// Return a random u64.
+pub fn rand_u64() -> Result<u64, ErrorStack> {
```
Comment (glorv) [inline components/crypto/src/rand.rs:13]: It's a bit weird that get random number can fail? Can we fallback to system rand if this call failed?

Comment (Connor1996) [reply components/crypto/src/rand.rs:13]: +1, can we just unwrap it

Comment (overvenus) [reply components/crypto/src/rand.rs:13]: In the context of cryptographic random number generation, RNG may fails due to lack of entropy.
And FIPS 140 has requirements to RNG, the default system RNG may not meet the requirements. 
So we can not simply unwrap the error.

**Review by glorv — APPROVED** [review]

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**